### PR TITLE
During particle advection only check for periodic boundaries in boundary cells

### PR DESCRIPTION
**Pull request by gassmoeller**
An optimization to the particle advection algorithm if the geometry has periodic boundaries. Instead of checking if a particle has crossed a periodic boundary in every cell, we should only check this in boundaries that actually have a periodic boundary face (we limit the CFL number to 1.0 for particle models anyway, so no particle that is not in a boundary cell should be able to cross a boundary, we assume this in other places like the sorting algorithm already). This change saves about 10% of particle advection costs in models with periodic boundaries.

### CODE REVIEW

File: source/particle/integrator/euler.cc [hunk 1]
```diff
@@ -40,7 +40,13 @@ namespace aspect
                           "to the number of particles to advect. For some unknown reason they are different, "
                           "most likely something went wrong in the calling function."));
 
-        const bool geometry_has_periodic_boundary = (this->get_geometry_model().get_periodic_boundary_pairs().size() != 0);
+        const auto &cell = begin_particle->get_surrounding_cell();
+        bool at_periodic_boundary = false;
+        for (const auto &face_index: cell->face_indices())
+          if (cell->at_boundary(face_index))
+            if (cell->has_periodic_neighbor(face_index))
+              at_periodic_boundary = true;
```
Comment (bangerth) [inline source/particle/integrator/euler.cc:48]: I think you can do even better, first by testing whether the geometry has periodic boundaries before starting the loop, and second by using `break` in the inner `if`:
```suggestion
        const auto &cell = begin_particle->get_surrounding_cell();
        bool at_periodic_boundary = false;
        if (this->get_geometry_model().get_periodic_boundary_pairs().size() != 0)
          for (const auto &face_index: cell->face_indices())
            if (cell->at_boundary(face_index))
              if (cell->has_periodic_neighbor(face_index))
              {
                at_periodic_boundary = true;
                break;
              }
```

File: source/particle/integrator/euler.cc [hunk 2]
```diff
@@ -40,7 +40,13 @@ namespace aspect
                           "to the number of particles to advect. For some unknown reason they are different, "
                           "most likely something went wrong in the calling function."));
 
-        const bool geometry_has_periodic_boundary = (this->get_geometry_model().get_periodic_boundary_pairs().size() != 0);
+        const auto &cell = begin_particle->get_surrounding_cell();
```
Comment (bangerth) [inline source/particle/integrator/euler.cc:43]: I think I would prefer making `cell` not a reference:
```suggestion
        const auto cell = begin_particle->get_surrounding_cell();
```

File: source/particle/integrator/rk_2.cc [hunk 3]
```diff
@@ -65,7 +65,12 @@ namespace aspect
                           "to the number of particles to advect. For some unknown reason they are different, "
                           "most likely something went wrong in the calling function."));
 
-        const bool geometry_has_periodic_boundary = (this->get_geometry_model().get_periodic_boundary_pairs().size() != 0);
+        const auto &cell = begin_particle->get_surrounding_cell();
+        bool at_periodic_boundary = false;
+        for (const auto &face_index: cell->face_indices())
+          if (cell->at_boundary(face_index))
+            if (cell->has_periodic_neighbor(face_index))
+              at_periodic_boundary = true;
```
Comment (bangerth) [inline source/particle/integrator/rk_2.cc:73]: Same two sets of comments apply here too.

File: source/particle/integrator/rk_4.cc [hunk 4]
```diff
@@ -69,7 +69,12 @@ namespace aspect
                           "to the number of particles to advect. For some unknown reason they are different, "
                           "most likely something went wrong in the calling function."));
 
-        const bool geometry_has_periodic_boundary = (this->get_geometry_model().get_periodic_boundary_pairs().size() != 0);
+        const auto &cell = begin_particle->get_surrounding_cell();
+        bool at_periodic_boundary = false;
+        for (const auto &face_index: cell->face_indices())
+          if (cell->at_boundary(face_index))
+            if (cell->has_periodic_neighbor(face_index))
+              at_periodic_boundary = true;
```
Comment (bangerth) [inline source/particle/integrator/rk_4.cc:77]: And here

**Review by bangerth — CHANGES_REQUESTED** [review]

**Comment by gassmoeller** [comment]
You are right, first checking if the geometry model has periodic boundaries at all makes the most sense (and is fastest for models that do not have periodic boundaries). One of the problems with the prior state of the code was that `this->get_geometry_model().get_periodic_boundary_pairs()` is actually a quite expensive function, because it constructs the data structure every time it is called and returns it by-value (`std::set<std::pair<.....>>`) which for every call requires dynamic memory allocation. I first solved this by introducing a simpler function into the ASPECT geometry models, but then I found the deal.II function `Triangulation<dim>::get_periodic_face_map` which should give me exactly the same information without constructing any objects. I think I prefer the current version the most, but let me know which version you prefer.

**Pull request merged**